### PR TITLE
ZK-3770: source map produces a wrong line number

### PR DESCRIPTION
**Pull request by DevChu**

### CODE REVIEW

File: zk/src/org/zkoss/zk/ui/http/SourceMapManager.java [hunk 1]
```diff
@@ -30,163 +45,152 @@
...
-	 */
-	public List<SourceMapInfo> getSourceMapInfoList() {
-		return _sourceMapInfoList;
+		_jsContentMap = new LinkedHashMap<>();
+		_postJsContentMap = new LinkedHashMap<>();
 	}
 
 	/**
 	 * Returns the source mapping url
+	 *
```
Comment (DevChu) [inline zk/src/org/zkoss/zk/ui/http/SourceMapManager.java:66]: remove

File: zk/src/org/zkoss/zk/ui/http/SourceMapManager.java [hunk 2]
```diff
@@ -30,163 +45,152 @@
...
-	public void insertSourceMap(int index, String sourceMapContent, int jsLineCount, String sourcePath) {
-		if (sourceMapContent == null || sourceMapContent.length() == 0) //add Empty
-			sourceMapContent = getEmptySourceMap(jsLineCount);
-		if (index > -1) //if -1, append
-			_sourceMapInfoList.add(index, new SourceMapInfo(sourceMapContent, jsLineCount, sourcePath));
-		else
-			_sourceMapInfoList.add(new SourceMapInfo(sourceMapContent, jsLineCount, sourcePath));
+	public void appendJsContent(String... scripts) {
+		if (this._jsCursor == null) {
+			log.debug("(" + _name + ")" + "invalid _jsCursor");
```
Comment (DevChu) [inline zk/src/org/zkoss/zk/ui/http/SourceMapManager.java:105]: {} replacement

Comment (DevChu) [reply zk/src/org/zkoss/zk/ui/http/SourceMapManager.java:105]: -> checkJSCursor...

File: zk/src/org/zkoss/zk/ui/http/SourceMapManager.java [hunk 3]
```diff
@@ -30,163 +45,152 @@
...
-			//clear info
-			_sourceMapInfoList.clear();
-		} catch (Exception e) {
-			log.warn("Failed to parse source map file. " + e.getMessage());
+	public void clearJsCursor() {
+		this._jsCursor = null;
+	}
+
+	public void closeJsCursor(ByteArrayOutputStream out) {
+		String moduleFuncName = _jsCursor.getY();
```
Comment (DevChu) [inline zk/src/org/zkoss/zk/ui/http/SourceMapManager.java:124]: realpath

File: zk/src/org/zkoss/zk/ui/http/SourceMapManager.java [hunk 4]
```diff
@@ -30,163 +45,152 @@
...
+		List<SourceFile> jsSourceFiles = new LinkedList<>();
+		handleSourceFiles(session, jsSourceFiles, _jsContentMap);
+		String wpdFileName = "js/src/" + _name + ".wpd.src.js";
+		String wpdFilePath = _sourceRoot + wpdFileName;
+		String wpdFileContent = this._preScript + new String(data) + this._postScript;
+		SourceFile wpdSourceFile = _builder.buildFromCode(wpdFilePath, wpdFileContent);
+		session.setAttribute("/" + wpdFileName, wpdFileContent);
+		jsSourceFiles.add(wpdSourceFile);
+		handleSourceFiles(session, jsSourceFiles, _postJsContentMap);
+		com.google.javascript.jscomp.Compiler compiler = new com.google.javascript.jscomp.Compiler();
```
Comment (DevChu) [inline zk/src/org/zkoss/zk/ui/http/SourceMapManager.java:159]: remove package

File: zk/src/org/zkoss/zk/ui/http/WpdExtendlet.java [hunk 5]
```diff
@@ -352,85 +350,71 @@ private Object parse(RequestContext reqctx, InputStream is, String path) throws
...
-							write(out, "[zk._p.p.");
-							write(out, info[0]);
-							write(out, ",'");
-							write(out, info[1]);
-							write(out, "'];\n");
-							if (sourceMapManager != null)
-								sourceMapManager.appendEmptySourceMap(1);
+							appendJsContent(out, sourceMapManager, "[zk._p.p.", info[0], ",'", info[1], "'];\n");
 						} else {
-							moldInfos.put(uri, new String[] { wgtnm, mold });
```
Comment (DevChu) [inline zk/src/org/zkoss/zk/ui/http/WpdExtendlet.java:374]: restore

File: zk/src/org/zkoss/zk/ui/http/WpdExtendlet.java [hunk 6]
```diff
@@ -440,27 +424,31 @@ private Object parse(RequestContext reqctx, InputStream is, String path) throws
...
-			writeHost(wc, out, getWebApp());
-			if (sourceMapManager != null)
-				sourceMapManager.addNonResolvedSourceMap();
+			writeHost(wc, out, getWebApp(), reqctx, sourceMapManager);
 		} else {
-			if (sourceMapManager != null)
-				sourceMapManager.appendEmptySourceMap(1);
-			write(out, "\n}catch(error){console.error(error);}finally{zk.setLoaded(zk._p.n);}");
+			if (sourceMapManager == null)
+				write(out, "\n}catch(error){console.error(error);}finally{zk.setLoaded(zk._p.n);}");
```
Comment (DevChu) [inline zk/src/org/zkoss/zk/ui/http/WpdExtendlet.java:436]: extract method

File: zk/src/org/zkoss/zk/ui/http/WpdExtendlet.java [hunk 7]
```diff
@@ -508,11 +496,14 @@ private boolean isWpdContentRequired(String pkg, Element root) {
...
-	private void writeHost(WpdContent wc, ByteArrayOutputStream out, WebApp wapp) {
+	private void writeHost(WpdContent wc, ByteArrayOutputStream out, WebApp wapp, RequestContext reqctx, SourceMapManager sourceMapManager) throws ServletException, IOException {
 		if (wapp != null) {
 			final String[] pkgs = wapp.getConfiguration().getClientPackages();
 			if (pkgs.length > 0) {
-				move(wc, out);
+				if (sourceMapManager == null)
+					move(wc, out);
+				else
+					sourceMapManager.appendPostScript(new String(wc.toByteArray(reqctx)));
```
Comment (DevChu) [inline zk/src/org/zkoss/zk/ui/http/WpdExtendlet.java:506]: extract?

File: zk/src/org/zkoss/zk/ui/http/WpdExtendlet.java [hunk 8]
```diff
@@ -713,11 +687,13 @@ else if (WebApps.getFeature("pe"))
...
-		write(out, sb.toString());
-		if (sourceMapManager != null)
-			sourceMapManager.appendEmptySourceMap(1);
+				.append(Encodes.encodeURL(ctx, reqctx.request, reqctx.response, wapp.getResourceURI(false)))
+				.append("'").append("});");
+		String appInfo = sb.toString();
+		if (sourceMapManager == null)
+			write(out, appInfo);
+		else
+			sourceMapManager.appendPostScript(appInfo);
```
Comment (DevChu) [inline zk/src/org/zkoss/zk/ui/http/WpdExtendlet.java:696]: extract?

File: zweb/src/org/zkoss/web/util/resource/ClassWebResource.java [hunk 9]
```diff
@@ -632,43 +633,57 @@ private void web0(HttpServletRequest request, HttpServletResponse response, Stri
...
+						_cwc.getLocator());
+				is = getResourceAsStream(orgpi);
+				if (is != null)
+					pi = orgpi;
+			} else if (ext != null && ext.endsWith("src.js") && session.getAttribute("zk$sourcemapsupported") != null) {
+				String cachedSource = (String) session.getAttribute(pi);
+				if (cachedSource != null) {
+					data = cachedSource.getBytes();
+					resolved = true;
+					session.removeAttribute(pi); //remove to prevent memory leak
```
Comment (DevChu) [inline zweb/src/org/zkoss/web/util/resource/ClassWebResource.java:652]: put into static?

File: zk/src/org/zkoss/zk/ui/http/WpdExtendlet.java [hunk 10]
```diff
@@ -107,7 +107,7 @@ public void service(HttpServletRequest request, HttpServletResponse response, St
 			throws ServletException, IOException {
 		// handle sourcemapping URL
 		if (isDebugJS() && path.endsWith("wpd"))
-			request.getSession().setAttribute(HANDLE_SOURCE_MAPPING_URL, true);
+			request.setAttribute(HANDLE_SOURCE_MAPPING_URL, true);
```
Comment (DevChu) [inline zk/src/org/zkoss/zk/ui/http/WpdExtendlet.java:110]: CLOSURE_COMPILER_AVAILABLE = Classes.existsByThread("com.google.debugging.sourcemap.SourceMapGeneratorV3");
->
com.google.javascript.jscomp.Compiler

File: zweb/src/org/zkoss/web/util/resource/ClassWebResource.java [hunk 11]
```diff
@@ -554,7 +564,7 @@ public void setDebugJS(boolean debug) {
 
 	/** Called by WebManager#setUpdateUri when WebManager is created
 	 * by HttpSessionListener#contextInitialized
-	 *
```
Comment (DevChu) [inline zweb/src/org/zkoss/web/util/resource/ClassWebResource.java:557]: remove

**Review by plijyfes — APPROVED** [review]

**Comment by scribetw** [comment]
@zkadmin rebase

**Comment by zkadmin** [comment]
I just pushed up the changes, enjoy!

**Pull request merged**